src/coreclr/vm/object.cpp

```diff
@@ -1545,13 +1558,11 @@ void StackTraceArray::CopyFrom(StackTraceArray const & src)
     }
     CONTRACTL_END;
 
-    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Capacity()));
-
-    Volatile<size_t> size = src.Size();
+    Volatile<uint32_t> size = src.Size();
     memcpyNoGCRefs(GetRaw(), src.GetRaw(), size * sizeof(StackTraceElement) + sizeof(ArrayHeader));
-
     SetSize(size);  // set size to the exact value which was used when we copied the data
                     // another thread might have changed it at the time of copying
+    SetKeepAliveItemsCount(ComputeKeepAliveItemsCount());
```

Comment:
I do not see why we are recomputing the keep alive items count here.

Reply:
The source stack trace might have new entries added. We've copied just "size" entries, but the keep alive items count may cover some of the entries that were added between the time we've read the size and copied the header and elements. So, we don't really know the actually keep alive count matching the number of elements we've copied, thus we compute it.

Reply:
Is this needed for all callers (there are not that many) or just `ExceptionObject::GetStackTrace`? It seems to be only needed for `ExceptionObject::GetStackTrace`. `ExceptionObject::GetStackTrace` does other consistency checks and we may just fold it into those.

Reply:
I have been trying to identify the operations that are performed on the stacktrace data structure by the rest of the system and review them for thread-safety. It is kind of hard since the operations are not clearly delineated in the code. It would really help if the operations performed on the stacktrace data structure were easy to identify.

 I am thinking about something like:
```
struct StackTrace
{
public:
    // Initialize with OBJECTREF
    StackTrace(OBJECTREF obj);
    // Extract OBJECTREF
    OBJECTREF GetValue();

    // Ensures that the current thread has exclusive write access to the stacktrace and there is space for given number of 
    // extra elements. Creates copy as necessary.
    // (THROWS, GC_TRIGGERS)
    void AcquireExlusiveAccess(int extraElements, int extraKeepAlive);

    // Appends the element to the stacktrace.  Assumes that AcquireExlusiveAccess has been called with extraElements/extraKeepAlive as appropriate.
    // (NOTHROW, GC_NOTRIGGER)
    void AppendStackTraceElement(const StackTraceElement& element, OBJECTREF keepAlive);

    // Freezes the stacktrace. Assumes that AcquireExlusiveAccess has been called.
    // (NOTHROW, GC_NOTRIGGER)
    void Freeze();

    // Get number of valid elements (cuts-off spurious extra elements introduced by race conditions)
    // (NOTHROW, GC_NOTRIGGER)
    int GetNumberOfStackTraceElements();

    // Return i-th stack element. i must be < GetNumberOfStackTraceElements().
    // (NOTHROW, GC_NOTRIGGER)
    StackTraceElement& GetStackTraceAt(int i);

private:
    ... private helper methods as necessary ...
};
```

Reply:
> Is this needed for all callers (there are not that many) or just `ExceptionObject::GetStackTrace`? It seems to be only needed for `ExceptionObject::GetStackTrace`. `ExceptionObject::GetStackTrace` does other consistency checks and we may just fold it into those.

There are just two callers - the `ExceptionObject::GetStackTrace` you've mentioned and the `StackTraceInfo::EnsureStackTraceArray` that always calls that with the stack trace that belongs to the current thread. I could change it so that there are two versions of `CopyDataFrom` / parametrized one where one does just copying of the array and the other called by the `ExceptionObject::GetStackTrace` would do the keep alive count recomputation.

As for the thread safety in general, the only place that needs to take care of stack traces coming from different thread and possibly mutating under our hands is the `ExceptionObject::GetStackTrace`.  So only this method (and the `CopyDataFrom` that's part of it) takes care of the dynamically changing stack trace / keepalive array. It always returns a stack trace that is owned by the current thread, so it can be safely iterated / modified. This is the key point of the new design. 

The only special case when the copy doesn't happen is in DAC.

Reply:
It does not look right that `CopyDataFrom` compensates for some inconsistencies introduced by race conditions, but not all of them. And that this compensation is actually only needed by one of the callers.

I think it would be best to get rid of the compensation for race conditions from `CopyDataFrom` and inline it into the `ExceptionObject::GetStackTrace` instead. It can be also more efficient that way. The existing loop in `ExceptionObject::GetStackTrace` can be used to prune both spurious StackTraceElements and spurious keepAlives.

Reply:
Done